src/artifactRegistry.ts

```diff
@@ -124,3 +205,103 @@ export class CachingDockerRegistryClient {
+ *   "version":"projects/platform-cross-environment/locations/us-central1/repositories/platform-docker/packages/identity/versions/sha256:123abc456defg"
+ * }
+ *
+ */
+export type DockerTag = {
+  name: string;
+  version: string;
+};
+
+/**
+ * getRelevantCommits
+ *
+ * @param {string} prevTag
+ *   The previous docker tag: of the following format: main---0013567-2024.04-g<githash>
+ * @param {string} nextTag
+ *  The next docker tag: main---0013567-2024.04-g<githash>
+ *
+ * @param {DockerTag[]} dockerTags
+ *  The docker tags as we need to parse and filter into relevant commits. Provided by a previous call to the artifact registry.
+ *
+ *
+ * @param {function} getTagFromDockerTag
+ *   Should map from `projects/platform-cross-environment/locations/us-central1/repositories/platform-docker/packages/servicename/tags/2022.02-278-g123456789` -> `2022.02-278-g123456789`
+ *   Should be a wrapper around the ArtifactRegistryClient, but written this way for testability, so the behavior can be injected.
+ *
+ * @returns {string[]} - The relevant commits as strings
+ */
+export function getRelevantCommits(
+  prevTag: string,
+  nextTag: string,
+  dockerTags: DockerTag[],
+  getTagFromDockerTag: (dockerTag: DockerTag) => string,
+): string[] {
+  if (!prevTag.startsWith('main')) return [];
+  if (!nextTag.startsWith('main')) return [];
+
+  /**
+   * Going to loop over our docker tags, filtering out ones outside the relevant range, and build an array of tag info.
+   *
+   * We will then do some commit deduping and return a list of relevant commits.
+   */
+  const relevantCommitsWithTagInfo = new Array<{
+    version: string;
+    tag: string;
+    commit: string;
+  }>();
+
+  for (const dockerTag of dockerTags) {
+    if (!dockerTag.version || !dockerTag.name) continue;
+
+    const tag = getTagFromDockerTag(dockerTag);
+
+    if (!tag.startsWith('main')) continue;
+    if (!tagInRange(prevTag, nextTag, tag)) continue;
+
+    // We only care about the tags between prev and next that have a git commit
+    const gitCommitMatches = tag.match(/-g([0-9a-fA-F]+)$/);
+    if (gitCommitMatches) {
+      const currTagInfo = {
+        version: dockerTag.version,
+        tag,
+        commit: gitCommitMatches[1],
+      };
+      relevantCommitsWithTagInfo.push(currTagInfo);
+    }
+  }
+
+  const result = dedupNeighboringTags(relevantCommitsWithTagInfo)
+    .sort((a, b) => a.tag.localeCompare(b.tag))
+    .map((c) => c.commit);
+
+  return result;
+}
+
+function tagInRange(prevTag: string, nextTag: string, tag: string): boolean {
+  return tag > prevTag && tag < nextTag;
+}
+
+function dedupNeighboringTags(
+  tags: { version: string; tag: string; commit: string }[],
+): { version: string; tag: string; commit: string }[] {
+  if (tags.length === 0) {
+    return [];
+  }
+
+  const res = [tags[0]];
+  for (let i = 1; i < tags.length; i++) {
+    const currTag = tags[i];
+    const prevTag = tags[i - 1];
+    if (currTag.commit !== prevTag.commit) {
```

Comment:
I think what we need to be de-duping is not just the commit, but the underlying Docker image SHA. The idea is that we are doing a build of every image on every push to main including a new tag, but due to reproducible builds most of the time we're just adding a new tag to an existing version (Docker image SHA). So we only want to keep around the tags (commits) that change the version (built image).